applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c

```diff
@@ -0,0 +1,786 @@
+static int do_carrier_device_power_sources(void)
+{
+	int ret;
+	uint8_t sources[7], count;
+	uint16_t source;
+
+	count = at_params_valid_count_get(&at_param_list);
+	if (count - 2 > sizeof(sources)) {
+		LOG_DBG("AT#XCARRIER=\"power_sources\" failed: too many parameters");
+		return -EINVAL;
+	}
+
+	for (int i = 2; i < count; i++) {
+		ret = at_params_unsigned_short_get(&at_param_list, i, &source);
+		if (ret) {
+			return ret;
+		}
+
+		sources[i - 2] = (uint8_t)source;
+	}
+
+	return lwm2m_carrier_avail_power_sources_set(sources, count - 2);
+}
+
+/* AT#XCARRIER="timezone","read|write"[,<timezone>] */
+static int do_carrier_device_timezone(void)
+{
+	int ret;
+	char operation[6];
+	size_t size = sizeof(operation);
+
+	ret = util_string_get(&at_param_list, 2, operation, &size);
+	if (ret) {
+		return ret;
+	}
+
+	if (slm_util_cmd_casecmp(operation, "READ")) {
+		const char *timezone;
+
+		timezone = lwm2m_carrier_timezone_read();
+
+		if (timezone) {
+			sprintf(rsp_buf, "\r\n#XCARRIER: %s\r\n", timezone);
+		} else {
+			sprintf(rsp_buf, "\r\n#XCARRIER: \r\n");
+		}
+
+		rsp_send(rsp_buf, strlen(rsp_buf));
+
+		return 0;
+	} else if (slm_util_cmd_casecmp(operation, "WRITE")) {
+		char timezone[64];
+
+		size = sizeof(timezone);
+
+		ret = util_string_get(&at_param_list, 3, timezone, &size);
+		if (ret) {
+			return ret;
+		}
+
+		return lwm2m_carrier_timezone_write(timezone);
+	}
+
+	LOG_DBG("AT#XCARRIER=\"timezone\" failed: invalid operation");
+
+	return -EINVAL;
+}
+
+#define TIME_STR_SIZE sizeof("1970-01-01T00:00:00Z")
+
+static void print_utc_time(char *output, int32_t timestamp)
+{
+	time_t time = (time_t)timestamp;
+	struct tm *tm = gmtime(&time);
+
+	strftime(output, TIME_STR_SIZE, "%FT%TZ", tm);
+}
+
+/* AT#XCARRIER="time" */
+static int do_carrier_device_time(void)
+{
+	int utc_time, utc_offset;
+	const char *timezone = NULL;
+	char time_str[TIME_STR_SIZE];
+
+	lwm2m_carrier_time_read(&utc_time, &utc_offset, &timezone);
+
+	print_utc_time(time_str, utc_time);
+
+	if (timezone) {
+		sprintf(rsp_buf, "\r\n#XCARRIER: UTC_TIME: %s, UTC_OFFSET: %d, TIMEZONE: %s\r\n",
+		time_str, utc_offset, timezone);
+	} else {
+		sprintf(rsp_buf, "\r\n#XCARRIER: UTC_TIME: %s, UTC_OFFSET: %d, TIMEZONE:\r\n",
+		time_str, utc_offset);
+	}
+
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+	return 0;
+}
+
+/* AT#XCARRIER="utc_offset","read|write"[,<utc_offset>] */
+static int do_carrier_device_utc_offset(void)
+{
+	int ret, utc_offset;
+	char operation[6];
+	size_t size = sizeof(operation);
+
+	ret = util_string_get(&at_param_list, 2, operation, &size);
+	if (ret) {
+		return ret;
+	}
+
+	if (slm_util_cmd_casecmp(operation, "READ")) {
+		utc_offset = lwm2m_carrier_utc_offset_read();
+
+		sprintf(rsp_buf, "\r\n#XCARRIER: %d\r\n", utc_offset);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+
+		return 0;
+	} else if (slm_util_cmd_casecmp(operation, "WRITE")) {
+		ret = at_params_int_get(&at_param_list, 3, &utc_offset);
```

Comment:
I wonder if utc_offset points to a string here, not an int? I did not get the write to work

Reply:
It does point to an int. I had no problems getting this command to work as intended. What value did you try to write here?

Reply:
yes this is working in the test now.